### DISCUSSION TITLE
Ignoring files matching a pattern?

### DISCUSSION THREAD

**Discussion by omentic** [BurntSushi/ripgrep#2156] [Q&A] [answered]
Hi, I've been trying unsuccessfully to get ripgrep to ignore some files while searching through my hard drive (in this case, files under `/timeshift/`). What is the syntax to do this?

I think I'm missing something obvious - but I've tried everything I could find in the manual.

- `rg search / --iglob /timeshift/*`
- `rg search / --ignore-file ignore` (where ignore is a file containing `/timeshift/*`)
- `rg search /` (with a `.rgignore` file containing `/timeshift/*` in the home directory)
- `rg search /` (with a `.ignore` file containing `/timeshift/*` in the home directory)

**Comment by BurntSushi** [comment, marked as answer]
First I'll just comment directly on each thing you attempted, then I'll try to explain the unfortunate state of affairs today.

> `rg search / --iglob /timeshift/*`

So if you're trying to ignore things in `/timeshift`, then this definitely won't work because this is specifically asking ripgrep to _only_ search things inside of `/timeshift`. In order to negate a glob, you need to use `!`. So, in this case, `--iglob '!/timeshift/*'`. Note that the single quotes are important! (However, this probably still doesn't work for you.)

> `rg search / --ignore-file ignore` (where ignore is a file containing `/timeshift/*`)

This is _almost_ right, but as the docs say, `--ignore-file` is applied **relative to the current working directory**. So if your CWD isn't the root directory itself, then a rule like `/timeshift/*` will be interpreted as, "don't searching anything matching `timeshift/*`, but only where `timeshift` is an immediate child of the CWD." This is a gitignore rule. Namely, providing a leading `/` does _not_ mean "absolute path," but rather, acts as an anchor relative to where the ignore file is applied.

> `rg search /` (with a `.rgignore` file containing `/timeshift/*` in the home directory)

There should be zero docs suggesting that this would work how you expect it to.

> `rg search /` (with a `.ignore` file containing `/timeshift/*` in the home directory)

Same as above.

-----

OK, so the main problem here is that ripgrep uses gitignore match semantics everywhere, including in the `-g/--glob` flags. So when you use `/` in a glob pattern, it doesn't necessarily refer to an absolute path, but rather, refers to an _anchored path_ relative to where the globs are applied. So if your CWD is the root directory itself, then `-g '!/timeshift/*'` should do the trick for you. But if your CWD isn't the root directory, then that glob isn't going to work because it's only going to apply to `$CWD/timeshift` because of the leading slash. Now, you could use `-g '!timeshift/*'` (or even simpler, `-g '!timeshift'`), and that will work, but it may wind up ignoring more than you intend.

If changing your CWD isn't something you want to do, then I believe your only other choice is to drop a `.ignore` (or a `.rgignore`) at `/.ignore`, and put `/timeshift` there. That should work regardless of your CWD when you search `/`. And indeed, I just tried it and it works.

Now ideally, the `-g/--glob` flag would be re-worked a bit here. Arguably, it should be applied relative to each directory ripgrep traverses based on CLI args. I believe there are some open issues about this. It's a gnarly knot to untie though unfortunately.

**Reply by seriiserii825** [reply]
Hi. 
And how to exclude relative directories in config file?

This doesn't work.

```
-g
  --glob
    !/git/*
    !/autoload/*

```

**Reply by BurntSushi** [reply]
Please start a new question..... And please include full details, including all inputs, actual output and expected output.

Also, please consider reading the config file doc again. Your config file looks malformed to me.

**Reply by urev** [reply]
The 'single quote' matters:

`rg -g '!compile_commands.josn'` works.
`rg -g !compile_commands.json` not work.
